### PR TITLE
ppx: add [@drop_default] for record fields

### PR DESCRIPTION
**Pull request by andreypopp**
For now it only works for record fields annotated with `[@option]`, by dropping the field from JSON repr when the record value is `None`.

What's missing is to also support `[@drop_default]` for fields annotated by `[@default X]` but we need to decide how to check for equality between the default value and the field value.

One nice idea I had is to generate code like this:

```ocaml
type t = { a : int [@default 0] [@drop_default] }
...
let bnds =
  match [%equal int] a 0 with
  | true -> bnds
  | false -> ("a", a)::bnds
in
...
```

but this means this ppx will depend on another ppx which provides `[%equal t]` deriver but sadly `ppx_compare` doesn't work with melange now.

### CODE REVIEW

**Comment by anmonteiro** [comment]
> What's missing is to also support `[@drop_default]` for fields annotated by `[@default X]` but we need to decide how to check for equality between the default value and the field value.

I honestly don't think this is necessary at all, or at least I can't see a use case. Do you have an example where you'd want to use `[@drop_default]` but not an option type?

**Comment by andreypopp** [comment]
> I honestly don't think this is necessary at all, or at least I can't see a use case. Do you have an example where you'd want to use [@drop_default] but not an option type?


If your field type is not `_ option`, e.g. ` : int [@default 0]`, ` : _ list [@default []]` and so on. I think it's useful and atd and ppx_conv_yojson provide similar mechanisms.

But I agree that dropping the `None` is probably the most often used, hence I suggest to add this feature only for `[@option]` and then later see if we want to extend it further.

File: ppx/browser/ppx_deriving_json_js.ml [hunk 1]
```diff
@@ -190,10 +190,19 @@ module To_json = struct
     let loc = t.rcd_loc in
     let fs =
       List.map2 t.rcd_fields es ~f:(fun ld x ->
-          let n = ld.pld_name in
-          let n = Option.value ~default:n (ld_attr_json_key ld) in
-          let this = derive ld.pld_type x in
-          map_loc lident n, this)
+          let k = ld.pld_name in
+          let k = Option.value ~default:k (ld_attr_json_key ld) in
```
Comment (anmonteiro) [inline ppx/browser/ppx_deriving_json_js.ml:194]:
```suggestion
          let k =
            let k = ld.pld_name in
            Option.value ~default:k (ld_attr_json_key ld) in
```

Comment (anmonteiro) [reply ppx/browser/ppx_deriving_json_js.ml:194]: I kinda like this pattern, could also use it for `v`, but this is optional if you don't like the extra nesting

**Review by anmonteiro — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -249,6 +249,22 @@ let t = of_json (Json.parseOrRaise {|{"a": 42}|})
 (* t = { a = 42; b = None; } *)
 ```
 
+#### `[@json.drop_default]`: drop default values from JSON
+
+When a field has `[@option]` attribute one can use `[@json.drop_default]`
```
Comment (jchavarri) [inline README.md:254]: What happens now if we use `drop_default` in a non option field? Also, would it be worth to test this case? To have it documented "internally".

File: README.md [hunk 3]
```diff
@@ -249,6 +249,22 @@ let t = of_json (Json.parseOrRaise {|{"a": 42}|})
...
+attribute to make the generated `to_json` function to drop the field if it's
+value is `None`:
+
+```ocaml
+type t = {
+  a: int;
+  b: string option [@json.option] [@json.drop_default];
```
Comment (jchavarri) [inline README.md:261]: Is it `option` (line 254) or `json.option`?

**Review by jchavarri — COMMENTED** [review]
Could you add a changelog entry please?

I think I'd design the API differently to avoid corner cases and make it less verbose. We could do `[@option drop_default]` now, and in the future potentially `[@default 0, drop]`.

File: README.md (hunk 3 above)
Comment (andreypopp) [reply README.md:261]: It’s both, in ppxlib you register attributes fully qualified but they can be used both ways.

File: README.md (hunk 2 above)
Comment (andreypopp) [reply README.md:254]: added a test case

**Comment by andreypopp** [comment]
> Could you add a changelog entry please?

done

> I think I'd design the API differently to avoid corner cases and make it less verbose. We could do [@option drop_default] now, and in the future potentially [@default 0, drop].

`[@default 0, drop]` is problematic as `0, drop` is a tuple expression and can be a default value

what can be done is to make dropping the defaults the default behaviours, and instead add `[@keep_default]` to make it emit JSON with default values

this is probably ok for `[@option]` but for `[@default E]` when we add it, it means we'll need to require equality function defined for type of `E`

**Review by jchavarri — APPROVED** [review]
Thanks!
